zc_events/backends.py

```diff
@@ -0,0 +1,285 @@
+    It is also required that the consumers of the events on Rabbitmq be using celery 3 or 4.
+
+    Note:
+        It is not intended to be used directly by developers, but instead set and instantiated
+        through the RPC_BACKEND setting.
+
+        All public methods are backend implementations for the corresponding methods on EventClient,
+        except for the respond method.
+
+    Args:
+        redis_client (redis connection, (optional)): If this option is not provided, a redis connection
+            is gotten by using the EVENTS_REDIS_URL in your settings, using db 0.
+        pick_pool(pika_pool.QueuedPool, (optional)): The connection used by rabbitmq. If it is not provided,
+            a connection is established using the BROKER_URL from settings.
+    """
+
+    def __init__(self, redis_client=None, pika_pool=None):
+        if redis_client:
+            self._redis_client = redis_client
+        else:
+            pool = redis.ConnectionPool().from_url(settings.EVENTS_REDIS_URL, db=0)
+            self._redis_client = redis.Redis(connection_pool=pool)
+
+        if pika_pool:
+            self._pika_pool = pika_pool
+        else:
+            pika_params = pika.URLParameters(settings.BROKER_URL)
+            pika_params.socket_timeout = 5
+            self._pika_pool = pika_pool_lib.QueuedPool(
+                create=lambda: pika.BlockingConnection(parameters=pika_params),
+                max_size=10,
+                max_overflow=10,
+                timeout=10,
+                recycle=3600,
+                stale=45,
+            )
+
+        self._events_exchange = settings.EVENTS_EXCHANGE
+
+    def call(self, key, data):
+        return self.post(key, data)
+
+    def call_no_wait(self, key, data):
+        return self.post_no_wait(key, data)
+
+    def get(self, key, data):
+        data = _format_data(data, 'GET', key)
+        return self._enqueue_with_waiting(data)
+
+    def put(self, key, data):
+        data = _format_data(data, 'PUT', key)
+        return self._enqueue_with_waiting(data)
+
+    def put_no_wait(self, key, data):
+        data = _format_data(data, 'PUT', key)
+        return self._enqueue_without_waiting(data)
+
+    def post(self, key, data):
+        data = _format_data(data, 'POST', key)
+        return self._enqueue_with_waiting(data)
+
+    def post_no_wait(self, key, data):
+        data = _format_data(data, 'POST', key)
+        return self._enqueue_without_waiting(data)
+
+    def delete(self, key, data):
+        data = _format_data(data, 'DELETE', key)
+        return self._enqueue_with_waiting(data)
+
+    def delete_no_wait(self, key, data):
+        data = _format_data(data, 'DELETE', key)
+        return self._enqueue_without_waiting(data)
+
+    def _enqueue_without_waiting(self, data):
+        _place_on_queue(
+            self._pika_pool,
+            self._events_exchange,
+            _DEFAULT_ROUTING_KEY,
+            _LOW_PRIORITY,
+            json.dumps(data)
```

Comment:
Very minor, but I think `ujson` is faster than the stdlib `json` module for things like this so I'd suggest swapping it out everywhere we're using `json`. (That at least was the case last time we took a look at the benchmarks)

Reply:
✔️ I imported `ujson` as `json` since it's pretty much a drop in replacement (I know of no incompatibilities).